Automattic/wp-calypso
# Decouple CTA actions from plans-grid-next package.

Related to Automattic/martech#1458

## Proposed Changes
This PR is the first step to fully decouple the CTA configuration of the plans grid, inspired by @jdc91 's post pdvytD-qk-p2. Conceptually, it's an extension to the current `PlanActionOverrides` system, but it only replaces the callback functions part for now as the 1st step. 
The ultimate goal is that the `plans-grid-next` package simply offers the `PlanActions` prop for the calling site to configure the clicked actions, text, disabled or not, etc. The package itself shouldn't need to know those contextual info.

Summarizing the introduced changes:

* Introduce a new type, `PlanActions`, which is essentially a map encapsulating all the CTA info. Currently it only contains the on-click callback.
* Consolidate and migrate all the CTA functons into `use-plan-action` hooks. Note that the upgrade handler can obviously be simplified, but I keep the similar structure for now to not introduce too much change at once.
* Introduce an action handler mechanism, i.e. `PlanActionCallback`, which allows the calling site to introduce further actions on the CTA click. It also allows returning a boolean value to indicate whether it wants to trigger early return, so the actual shopping cart processing won't happen. Two cases supported by this are the domain/plan upsell modal and the domain upsell from home modal.
* Deprecate the current CTA callback props, e.g. onUpgradeClick, handleUpgradeClick, `callback` in `PlanActionOverrides`, etc.
* The actual on-click callback has been separated finer, e.g. managing add-ons, managing plans, redirecting to the VIP page, handling uprades, instead of being combined into one like it currently does. 


TODO:

- [ ] Fix WooExpress case
- [ ] Fix https://github.com/Automattic/wp-calypso/blob/trunk/packages/plans-grid-next/src/components/test/actions.ts

## Follow-up TODOs

* Re-implement the downgrading behavior of the previous attempt: https://github.com/Automattic/wp-calypso/pull/84128 
* Migrate CTA text configuration under the same `PlanActions` mechanism introduced here.
* It's questionable that `isFreeTrial` is part of the final callback, but I'm not entirely sure what's the better way forward for now. Conceptually speaking, each grid cell simpy provides two CTAs: a primary one and a "hidden one"(it's only available on the Creator free trial at the moment), so maybe a simpler way is to just offer two CTA callback slots.
* `usePlanActions` should have its own unit tests, to make sure it assign the right CTA function given certain inputs.
* Simplify the upgrade handler.

## Testing Instructions



* Purchase a plan with/without the storage add-on on a classic sign-up flow, e.g. /start/, should work as expected.
* Purchase a plan with/without the storage add-on on a Stepper flow, e.g. /setup/newsletter, should work as expected.
* Purchase a plan with/without the storage add-on on /plans should work as expected.
* P2 flow should work as expceted.
* WooExpress flow should work as expected.
* The domain/plan upsell modal should work as expected in both `/start` and `/start/onboarding-pm`. 
    * Pick a free domain and then the free plan in /start/onboarding-pm
    * Pick a paid domain and then the free plan in both /start and /start/onboarding-pm 
* Domain upsell from home modal: 
    * On a Free site, access `/plans/{site slug}?get_domain=a-random-domain.xxx`
    * Click "Keep plan", and make sure the modal appear and works as expected.
* The Creator trial:
    * Enable the trial by hard coding like so: 334e8-pb
    * In /start/hosting, make sure both "Start Free Trial" and "Get Creator" work as expected.

## Pre-merge Checklist
